File: ControlAssuranceAPI/Models/CLCase.cs
using System;
using System.Collections.Generic;

namespace CAT.Models;

public partial class CLCase
{
    public int ID { get; set; }

    public string? Title { get; set; }
    
    public string? DeptTransferringTo { get; set; }
    
    public string? CaseType { get; set; }

    public int? CreatedById { get; set; }

    public DateTime? CreatedOn { get; set; }

    public int? ApplHMUserId { get; set; }

    public string? ReqCostCentre { get; set; }

    public int? ReqDirectorateId { get; set; }

    public string? ReqVacancyTitle { get; set; }

    public int? ReqGradeId { get; set; }

    public string? ReqWorkPurpose { get; set; }

    public int? ReqProfessionalCatId { get; set; }

    public DateTime? ReqEstStartDate { get; set; }

    public DateTime? ReqEstEndDate { get; set; }

    public int? ReqWorkLocationId { get; set; }

    public int? ReqNumPositions { get; set; }

    public int? ComFrameworkId { get; set; }

    public string? ComJustification { get; set; }

    public string? ComPSRAccountId { get; set; }

    public string? JustAltOptions { get; set; }

    public string? JustSuccessionPlanning { get; set; }

    public decimal? FinMaxRate { get; set; }

    public decimal? FinEstCost { get; set; }

    public int? FinIR35ScopeId { get; set; }

    public int? FinIR35AssessmentId { get; set; }

    public string? OtherComments { get; set; }

    public int? BHUserId { get; set; }

    public int? FBPUserId { get; set; }

    public int? HRBPUserId { get; set; }

    public string? BHApprovalDecision { get; set; }

    public string? BHApprovalComments { get; set; }

    public string? FBPApprovalDecision { get; set; }

    public string? FBPApprovalComments { get; set; }

    public string? HRBPApprovalDecision { get; set; }

    public string? HRBPApprovalComments { get; set; }

    public string? CaseChangeLog { get; set; }

    public bool? CaseCreated { get; set; }

    public int? CaseRef { get; set; }

    public int? BHDecisionById { get; set; }

    public int? FBPDecisionById { get; set; }

    public int? HRBPDecisionById { get; set; }

    public DateTime? BHDecisionDate { get; set; }

    public DateTime? FBPDecisionDate { get; set; }

    public DateTime? HRBPDecisionDate { get; set; }

    public string? FinApproachAgreeingRate { get; set; }

    public string? FinSummaryIR35Just { get; set; }

    public string? CLApprovalDecision { get; set; }

    public int? CLDecisionById { get; set; }

    public DateTime? CLDecisionDate { get; set; }
    public int? CBPUserId { get; set; }
    public string? CBPApprovalDecision { get; set; }
    public int? CBPDecisionById { get; set; }
    public System.DateTime? CBPDecisionDate { get; set; }
    public decimal? FinBillableRate { get; set; }
    public int? FinTotalDays { get; set; }
    public string? FinCalcType { get; set; }
    public decimal? FinCostPerWorker { get; set; }

    public virtual ICollection<CLHiringMember> CLHiringMembers { get; } = new List<CLHiringMember>();

    public virtual ICollection<CLWorker> CLWorkers { get; } = new List<CLWorker>();

    public virtual CLComFramework? ComFramework { get; set; }

    public virtual CLIR35Scope? CLIR35Scope { get; set; }

    public virtual Directorate? Directorate { get; set; }

    public virtual CLStaffGrade? CLStaffGrade { get; set; }

    public virtual CLProfessionalCat? CLProfessionalCat { get; set; }

    public virtual CLWorkLocation? CLWorkLocation { get; set; }
}

public static class ApprovalDecisions
{
    public readonly static string Approve = "Approve";
    public readonly static string Reject = "Reject";
    public readonly static string RequireDetails = "RequireDetails";
}

public static class CaseStages
{
    public readonly static CaseStage Draft = new CaseStage { Name = "Draft", Number = 1 };
    public readonly static CaseStage Approval = new CaseStage { Name = "Approval", Number = 2 };
    public readonly static CaseStage Onboarding = new CaseStage { Name = "Onboarding", Number = 3 };
    public readonly static CaseStage Engaged = new CaseStage { Name = "Engaged", Number = 4 };
    public readonly static CaseStage Leaving = new CaseStage { Name = "Leaving", Number = 5 };
    public readonly static CaseStage Left = new CaseStage { Name = "Left", Number = 6 };
    public readonly static CaseStage Extended = new CaseStage { Name = "Extended", Number = 7 };

    public static int GetStageNumber(string stageName)
    {
        if (stageName == CaseStages.Draft.Name)
            return CaseStages.Draft.Number;
        else if (stageName == CaseStages.Approval.Name)
            return CaseStages.Approval.Number;
        else if (stageName == CaseStages.Onboarding.Name)
            return CaseStages.Onboarding.Number;
        else if (stageName == CaseStages.Engaged.Name)
            return CaseStages.Engaged.Number;
        else if (stageName == CaseStages.Leaving.Name)
            return CaseStages.Leaving.Number;
        else if (stageName == CaseStages.Left.Name)
            return CaseStages.Left.Number;
        else if (stageName == CaseStages.Extended.Name)
            return CaseStages.Extended.Number;
        else
            return 0;
    }


    public class CaseStage
    {
        public string? Name { get; set; }
        public int Number { get; set; }
    }

    

}

public class ClCaseInfoView_Result
{
    public int ID { get; set; }
    public string? Title { get; set; }
    public string? Stage { get; set; }
    public string? CreatedBy { get; set; }
    public string? CreatedOn { get; set; }
    public string? CaseRef { get; set; }
    public string? ApplHMUser { get; set; }
    public string? ApplHMembers { get; set; }
    public string? ReqGrade { get; set; }
    public string? Directorate { get; set; }
    public string? ReqEstStartDate { get; set; }
    public string? ReqEstEndDate { get; set; }
    public string? ReqProfessionalCat { get; set; }
    public string? ReqWorkLocation { get; set; }
    public string? ComFramework { get; set; }
    public string? ComPSRAccount { get; set; }
    public string? FinIR35Scope { get; set; }
    public string? BHUser { get; set; }
    public string? FBPUser { get; set; }
    public string? HRBPUser { get; set; }
    public string? CBPUser { get; set; }
    public string? BHDecisionByAndDate { get; set; }
    public string? FBPDecisionByAndDate { get; set; }
    public string? HRBPDecisionByAndDate { get; set; }
    public string? CBPDecisionByAndDate { get; set; }
    public string? CLDecisionByAndDate { get; set; }
    public string? OnbContractorGender { get; set; }
    public string? OnbContractorTitle { get; set; }
    public string? OnbContractorDobStr { get; set; }
    public string? OnbStartDateStr { get; set; }
    public string? OnbEndDateStr { get; set; }
    public string? OnbSecurityClearance { get; set; }
    public string? WorkDays { get; set; }
    public string? OnbDecConflict { get; set; }
    public string? OnbLineManagerUser { get; set; }
    public string? OnbLineManagerGrade { get; set; }
    public string? OnbWorkOrderNumber { get; set; }
    public string? OnbRecruitersEmail { get; set; }
    public string? BPSSCheckedBy { get; set; }
    public string? BPSSCheckedOn { get; set; }
    public string? POCheckedBy { get; set; }
    public string? POCheckedOn { get; set; }
    public string? ITCheckedBy { get; set; }
    public string? ITCheckedOn { get; set; }
    public string? UKSBSCheckedBy { get; set; }
    public string? UKSBSCheckedOn { get; set; }
    public string? PassCheckedBy { get; set; }
    public string? PassCheckedOn { get; set; }
    public string? ContractCheckedBy { get; set; }
    public string? ContractCheckedOn { get; set; }
    public string? EngPONumber { get; set; }
    public string? EngPONote { get; set; }
    public string? SDSCheckedBy { get; set; }
    public string? SDSCheckedOn { get; set; }
    public string? SDSNotes { get; set; }
    public string? LeEndDateStr { get; set; }
    public string? LeContractorDetailsCheckedBy { get; set; }
    public string? LeContractorDetailsCheckedOn { get; set; }
    public string? LeITCheckedBy { get; set; }
    public string? LeITCheckedOn { get; set; }
    public string? LeUKSBSCheckedBy { get; set; }
    public string? LeUKSBSCheckedOn { get; set; }
    public string? LePassCheckedBy { get; set; }
    public string? LePassCheckedOn { get; set; }
    public string ExtensionHistory { get; set; } = "";
}

public class CLCaseView_Result
{
    public int ID { get; set; }
    public int CaseId { get; set; }
    public string? DeptTransferringTo { get; set; }
    public string? CaseRef { get; set; }
    public string? Title1 { get; set; }
    public string? Title2 { get; set; }
    public string? Stage { get; set; }
    public string? StageActions1 { get; set; }
    public string? StageActions2 { get; set; }
    public string? Worker { get; set; }
    public string? CreatedOn { get; set; }
    public string? CostCenter { get; set; }
    public string? HiringManager { get; set; }
    public int? HiringManagerId { get; set; }
    public string? EngagedChecksDone { get; set; }
    public string? StartDate { get; set; }
    public string? EndDate { get; set; }
}

public class CLCaseCounts_Result
{
    public int ID { get; set; } = 1;
    public int TotalBusinessCases { get; set; }
    public int TotalEngagedCases { get; set; }
    public int TotalArchivedCases { get; set; }
}

File: control-assurance-sp/src/types/CLCase.ts
import { IEntity } from "./Entity";


export interface ICLCase  extends IEntity {

    CaseType?: string;
    CreatedById?: number;
    CreatedOn?: Date;
    ApplHMUserId?: number;
    ReqCostCentre?: string;
    ReqDirectorateId?: number;
    ReqVacancyTitle?: string;
    ReqGradeId?: number;
    ReqWorkPurpose?: string;
    ReqProfessionalCatId?: number;
    ReqEstStartDate?: Date;
    ReqEstEndDate?: Date;
    ReqWorkLocationId?: number;
    ReqNumPositions?: number;
    ComFrameworkId?: number;
    ComJustification?: string;
    ComPSRAccountId?: string;
    JustAltOptions?: string;
    JustSuccessionPlanning?: string;
    FinMaxRate?: number;
    FinEstCost?: number;
    FinIR35ScopeId?: number;
    FinIR35AssessmentId?: number;    
    FinApproachAgreeingRate?: string;
    FinSummaryIR35Just?: string;
    OtherComments?: string;
    BHUserId?: number;
    FBPUserId?: number;
    HRBPUserId?: number;
    CBPUserId?: number;
    BHApprovalDecision?: string;
    BHApprovalComments?: string;
    FBPApprovalDecision?: string;
    FBPApprovalComments?: string;
    HRBPApprovalDecision?: string;
    HRBPApprovalComments?: string;
    CLApprovalDecision?: string;
    CBPApprovalDecision?: string;
    CaseChangeLog?: string;


}

export class CLCase implements ICLCase{ 
    public ID: number = 0;
    public Title: string = null;
    public CaseType?: string = null;
    public CreatedById?: number = null;
    public CreatedOn?: Date = null;
    public ApplHMUserId?: number = null;
    public ReqCostCentre?: string = null;
    public ReqDirectorateId?: number = null;
    public ReqVacancyTitle?: string = null;
    public ReqGradeId?: number = null;
    public ReqWorkPurpose?: string = null;
    public ReqProfessionalCatId?: number = null;
    public ReqEstStartDate?: Date = null;
    public ReqEstEndDate?: Date = null;
    public ReqWorkLocationId?: number = null;
    public ReqNumPositions?: number = 1;
    public ComFrameworkId?: number = null;
    public ComJustification?: string = null;
    public ComPSRAccountId?: string = null;
    public JustAltOptions?: string = null;
    public JustSuccessionPlanning?: string = null;
    public FinMaxRate?: number = null;
    public FinEstCost?: number = null;
    public FinIR35ScopeId?: number = null;
    public FinIR35AssessmentId?: number = null;
    public FinApproachAgreeingRate?: string = null;
    public FinSummaryIR35Just?: string = null;
    public OtherComments?: string = null;
    public BHUserId?: number = null;
    public FBPUserId?: number = null;
    public HRBPUserId?: number = null;
    public CBPUserId?: number = null;
    public BHApprovalDecision?: string = null;
    public BHApprovalComments?: string = null;
    public FBPApprovalDecision?: string = null;
    public FBPApprovalComments?: string = null;
    public HRBPApprovalDecision?: string = null;
    public HRBPApprovalComments?: string = null;
    public CLApprovalDecision?: string = null;
    public CBPApprovalDecision?: string = null;
    public CaseChangeLog?: string = null;

    constructor(caseType: string) {
        this.CaseType = caseType;

    }


}

export interface IClCaseInfo  extends IEntity {
    Stage?: string;
    CreatedBy?: string;
    CreatedOn?: string;
    CaseRef?: string;

    ApplHMUser?: string;
    ApplHMembers?: string;
    ReqGrade?: string;
    Directorate?: string;
    ReqEstStartDate?: string;
    ReqEstEndDate?: string;
    ReqProfessionalCat?: string;
    ReqWorkLocation?: string;
    ComFramework?: string;
    ComPSRAccount?: string;
    FinIR35Scope?: string;
    BHUser?: string;
    FBPUser?: string;
    HRBPUser?: string;
    BHDecisionByAndDate?: string;
    FBPDecisionByAndDate?: string;
    HRBPDecisionByAndDate?: string;
    CBPDecisionByAndDate?: string;
    CLDecisionByAndDate?: string;

    OnbContractorGender?: string;
    OnbContractorTitle?: string;
    OnbContractorDobStr?: string;
    OnbStartDateStr?: string;
    OnbEndDateStr?: string;
    OnbSecurityClearance?: string;
    WorkDays?: string;
    OnbDecConflict?: string;
    OnbLineManagerUser?: string;
    OnbLineManagerGrade?: string;
    OnbWorkOrderNumber?: string;
    OnbRecruitersEmail?: string;

    BPSSCheckedBy?: string;
    BPSSCheckedOn?: string;
    POCheckedBy?: string;
    POCheckedOn?: string;
    ITCheckedBy?: string;
    ITCheckedOn?: string;
    UKSBSCheckedBy?: string;
    UKSBSCheckedOn?: string;
    PassCheckedBy?: string;
    PassCheckedOn?: string;
    ContractCheckedBy?: string;
    ContractCheckedOn?: string;
    EngPONumber?:string;
    EngPONote?:string;
    SDSCheckedBy?:string;
    SDSCheckedOn?:string;
    SDSNotes?:string;


    LeEndDateStr?: string;
    LeContractorDetailsCheckedBy?: string;
    LeContractorDetailsCheckedOn?: string;
    LeITCheckedBy?: string;
    LeITCheckedOn?: string;
    LeUKSBSCheckedBy?: string;
    LeUKSBSCheckedOn?: string;
    LePassCheckedBy?: string;
    LePassCheckedOn?: string;

    ExtensionHistory?: string;

}
export class ClCaseInfo  implements IClCaseInfo {

    public ID: number = 0;
    public Title: string = "";
    public Stage?: string = "";
    public CreatedBy?: string = "";
    public CreatedOn?: string = "";
    public CaseRef?: string = "";

    public ApplHMUser?: string = "";
    public ReqGrade?: string = "";
    public Directorate?: string = "";
    public ReqEstStartDate?: string = "";
    public ReqEstEndDate?: string = "";
    public ReqProfessionalCat?: string = "";
    public ReqWorkLocation?: string = "";
    public ComFramework?: string = "";
    public ComPSRAccount?: string = "";
    public FinIR35Scope?: string = "";
    public BHUser?: string = "";
    public FBPUser?: string = "";
    public HRBPUser?: string = "";

    public BHDecisionByAndDate?: string = "";
    public FBPDecisionByAndDate?: string = "";
    public HRBPDecisionByAndDate?: string = "";

    public OnbContractorTitle?: string = "";
    public OnbContractorGender?: string = "";
    public OnbContractorDobStr?: string = "";
    public OnbStartDateStr?: string = "";
    public OnbEndDateStr?: string = "";
    public OnbSecurityClearance?: string = "";
    public WorkDays?: string = "";
    public OnbDecConflict?: string = "";
    public OnbLineManagerUser?: string = "";
    public OnbLineManagerGrade?: string = "";

    public BPSSCheckedBy?: string = "";
    public BPSSCheckedOn?: string = "";
    public POCheckedBy?: string = "";
    public POCheckedOn?: string = "";
    public ITCheckedBy?: string = "";
    public ITCheckedOn?: string = "";
    public UKSBSCheckedBy?: string = "";
    public UKSBSCheckedOn?: string = "";
    public PassCheckedBy?: string = "";
    public PassCheckedOn?: string = "";
    public ContractCheckedBy?: string = "";
    public ContractCheckedOn?: string = "";
    public SDSCheckedBy?:string = "";
    public SDSCheckedOn?:string = "";
    public SDSNotes?:string = "";

    public LeStartDateStr?: string = "";
    public LeContractorDetailsCheckedBy?: string = "";
    public LeContractorDetailsCheckedOn?: string = "";
    public LeITCheckedBy?: string = "";
    public LeITCheckedOn?: string = "";
    public LeUKSBSCheckedBy?: string = "";
    public LeUKSBSCheckedOn?: string = "";
    public LePassCheckedBy?: string = "";
    public LePassCheckedOn?: string = "";

    public ExtensionHistory = "";

}

export interface IClCaseCounts extends IEntity{
    TotalBusinessCases?: number;
    TotalEngagedCases?: number;
    TotalArchivedCases?: number;
}